src/main/java/de/tum/in/www1/artemis/web/rest/GenericSubmissionResource.java

```diff
@@ -0,0 +1,241 @@
+            return notFound();
+        }
+
+        // Check if the limit of simultaneously locked submissions has been reached
+        submissionService.checkSubmissionLockLimit(exercise.getCourse().getId());
+        return null;
+    }
+
+    /**
+     * Returns the data needed for the editor, which includes the participation, submission with answer if existing and the assessments if the submission was already
+     * submitted.
+     *
+     * @param participationId the id of the participation for which to find the data for the editor
+     * @param exerciseType type of the exercise for which we take the data
+     * @param submissionType type of the submission for which we take the data
+     * @param newSubmission new instance of concrete submission which can be needed if we don't find submission
+     * @return the ResponseEntity with the participation as body
+     */
+    final <E extends Exercise> ResponseEntity<T> getDataForEditor(long participationId, Class<E> exerciseType, Class<T> submissionType, T newSubmission) {
+        final StudentParticipation participation = participationService.findOneWithEagerSubmissionsAndResults(participationId);
+        if (participation == null) {
+            return ResponseEntity.notFound()
+                    .headers(HeaderUtil.createFailureAlert(applicationName, true, ENTITY_NAME, "participationNotFound", "No participation was found for the given ID.")).build();
+        }
+        E exercise;
+        if (exerciseType.isInstance(participation.getExercise())) {
+            exercise = exerciseType.cast(participation.getExercise());
+            if (exercise == null) {
+                return ResponseEntity.badRequest()
+                        .headers(HeaderUtil.createFailureAlert(applicationName, true, "exercise", "exerciseEmpty", "The exercise belonging to the participation is null."))
+                        .body(null);
+            }
+
+            // make sure sensitive information are not sent to the client
+            exercise.filterSensitiveInformation();
+        }
+        else {
+            return ResponseEntity.badRequest().headers(HeaderUtil.createFailureAlert(applicationName, true, "fileUploadExercise", "wrongExerciseType",
+                    "The exercise of the participation is not a file upload exercise.")).body(null);
+        }
+
+        // Students can only see their own file uploads (to prevent cheating). TAs, instructors and admins can see all file uploads.
+        if (!(authCheckService.isOwnerOfParticipation(participation) || authCheckService.isAtLeastTeachingAssistantForExercise(exercise))) {
+            return forbidden();
+        }
+
+        Optional<T> optionalSubmission = participation.findLatestSubmissionOfType(submissionType);
+        T submission;
+        if (optionalSubmission.isEmpty()) {
+            // this should never happen as the submission is initialized along with the participation when the exercise is started
```

Comment:
If this should really never happen, shouldn't we just throw an exception in these cases? Do we really instantiate a submission with the start of an exercise (for every exercise type) as the comment suggests?

Reply:
Yes we do. This is done when the student clicks Start Exercise on the course page. We do it for text, modeling and file upload exercises. The logic is in `startParticipation` method in `ParticipationResource`. But to be honest for me the logic there is quite complex and specifically handles each exercise type.
I would still not throw the exception for now as I don't trust the logic in `ParticipationService`. We can change it after the Partcipation part is refactored as well.